File: pin_draft.php
<?php

date_default_timezone_set('America/New_York');

/*

# YOU PROBABLY SHOULDN'T USE THIS. USE THE [PYTHON SCRIPT] INSTEAD

* Add `# # # # # php /path/to/script/pin_draft.php` to crontab.
* Script will loop through 18 times, which should take about a minute, then die
* Cron will run the script every minute so you should never have downtime

[PYTHON SCRIPT]: https://github.com/nickwynja/scripts/blob/master/pin_draft.py

*/

/* Pinboard.in Credentials and Tag */
$pinToken = '';
$pinTag = ''; //Drafts will only be created for pins with this tag. Can't be empty.

/* Draft location and extension */
$draftPath = '/path/to/drafts/';
$draftExt = '.md';

$pinAPI = 'api.pinboard.in/v1/';
$pinUpdate = 'posts/update';
$pinGet = 'posts/get';

$triggeredTime = 0;

function get_data($url) {
  $ch = curl_init();
  $timeout = 5;
  curl_setopt($ch,CURLOPT_URL,$url);
  curl_setopt($ch,CURLOPT_RETURNTRANSFER,1);
  curl_setopt($ch,CURLOPT_CONNECTTIMEOUT,$timeout);
  $data = curl_exec($ch);
  curl_close($ch);
  return $data;
}

function xml_attribute($object, $attribute){
  if(isset($object[$attribute]))
        return (string) $object[$attribute];
}

function slugify($text) {
  $text = preg_replace('~[^\\pL\d]+~u', '-', $text);
  $text = trim($text, '-');
  $text = iconv('utf-8', 'us-ascii//TRANSLIT', $text);
  $text = strtolower($text);
  $text = preg_replace('~[^-\w]+~', '', $text);

  return $text;
}

for ($i = 0; $i < 18; $i++) {

  $updateURL = 'https://' . $pinAPI . $pinUpdate . '?auth_token=' . $pinToken;
  $xml = simplexml_load_string(get_data($updateURL));
  $updateTime = strtotime(xml_attribute($xml, 'time'));

  if ($updateTime != $triggeredTime) {
   
    $getURL = 'https://' . $pinAPI . $pinGet . '?auth_token=' . $pinToken . '&tag=' . $pinTag . '&meta=yes';
    $xml = simplexml_load_string(get_data($getURL));

    if (property_exists($xml, 'post')) {
              
      $mostRecent = count($xml->post) - 1;
          
      $postURL = $xml->post[$mostRecent]->attributes()->href;
      $postSlug = slugify($xml->post[$mostRecent]->attributes()->description);
      $postTitle = $xml->post[$mostRecent]->attributes()->description;
      $postText = $xml->post[$mostRecent]->attributes()->extended;
      $meta = '';

      if (($xml->post[$mostRecent]->attributes()->meta) != $meta) {
        
        $draft = $postTitle . "\n====\nlink: " . $postURL . "  \npublish-not  \n\n" . $postText;
        $meta = $xml->post[$mostRecent]->attributes()->meta;    
        $file = $draftPath . $postSlug . $draftExt;
	$date = date('y-m-d H:i:s'); 

	if (!file_exists($file)) {
          $msg = file_put_contents($file, $draft);
          ($msg == FALSE) ? error_log('Draft write failed.') : error_log($postTitle . ' draft created at ' . $date);
	
	} else {

	  error_log($postTitle . ' draft already exists.');
	
	}      

	$triggeredTime = $updateTime;
    
      }

    } else {

      $triggeredTime = $updateTime;

    }
    
    sleep(3);
    
  } else {
    
    sleep(3);
 
  }
}
?>

File: pin_draft.py
#!/usr/bin/python

import json
import urllib2
import re
from unicodedata import normalize
import os, time
import datetime
import re
import logging

logging.basicConfig(format='%(asctime)s %(message)s', datefmt='%Y-%m-%d %H:%M:%S:',filename='/tmp/pin_draft.log',level=logging.INFO)
os.environ['TZ'] = 'America/New_York'
time.tzset()
now =  str(datetime.datetime.now())


# Pinboard.in Credentials and Tag

pinToken = 'nickwynja:41B828BD4C8B854841EE'
pinTag = 'hm'

# Draft location and extension
draftPath = '/home/blog/Dropbox/hackmake/drafts/'
draftExt = '.md'

pinAPI = 'api.pinboard.in/v1/'
pinUpdate = 'posts/update'
pinGet = 'posts/get'

triggeredTime = 0

def getJSON(url):
    req = urllib2.Request(url, None)
    opener = urllib2.build_opener()
    f = opener.open(req)
    return json.load(f)

_slugify_strip_re = re.compile(r'[^\w\s-]')
_slugify_hyphenate_re = re.compile(r'[-\s]+')
def slugify(value):
    import unicodedata
    if not isinstance(value, unicode):
        value = unicode(value)
    value = unicodedata.normalize('NFKD', value).encode('utf-8', 'ignore')
    value = unicode(_slugify_strip_re.sub('', value).strip().lower())
    return _slugify_hyphenate_re.sub('-', value)

############################################

for i in range(18):


    updateURL = 'https://' + pinAPI + pinUpdate + '?auth_token=' + pinToken + '&format=json'
    j = getJSON(updateURL)
    updateTime = j['update_time']
        
    if updateTime != triggeredTime:
    
        getURL = 'https://' + pinAPI + pinGet + '?auth_token=' + pinToken + '&tag=' + pinTag + '&meta=yes' + '&format=json'
        j = getJSON(getURL)
            
        if j['posts']:
    
            last = len(j['posts'])-1
            post = j['posts'][last]
            
            postURL = post['href']
            postSlug = slugify(post['description'])
            postTitle = post['description']
            postText = post['extended']
            metatmp = '/tmp/pin_draft.meta'
            with open(metatmp, 'r+') as f:
              meta = f.read()
              f.close()
           
            if post['meta'] != meta:
    
                draft = postTitle + "\n====\nlink: " + postURL + "  \npublish-not\n\n" + postText
                meta = post['meta']
                with open(metatmp, 'w') as f:
                  f.write(meta)
                  f.close()

                file = draftPath + postSlug + draftExt
                            
                if not os.path.exists(file):
                  try:
                    with open(file, 'w') as f:
                      f.write(draft.encode('utf-8', 'ignore'))
                      logging.info(postSlug + draftExt + ' draft created.')
                      f.closed
                  except EnvironmentError:
                    logging.info('Draft write failed.')
                
                triggeredTime = updateTime
            
        else:    
            triggeredTime = updateTime

        time.sleep(3)

    else:
      time.sleep(3)
